Shared informer custom resync periods

ncdc opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/kubernetes/blob/master/CONTRIBUTING.md and developer guide https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/faster_reviews.md
3. Follow the instructions for writing a release note: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/pull-requests.md#release-notes
-->

**What this PR does / why we need it**: add the ability for event handlers to specify their own distinct resync periods to a shared informer

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #26406 

**Special notes for your reviewer**:

This is part of my series of PRs to convert as much of the code base to using generated shared informers as possible. Assuming this merges, we'll also need #40385, and I'll probably close #40097 and replace it with a series of individual PRs (1 per controller change).

Assuming we get close to agreement on the implementation, I can look at adding more unit tests.

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```release-note
```

@smarterclayton @deads2k @liggitt @sttts @derekwaynecarr @pmorie @lavalamp @wojtek-t @gmarek @timothysc @jayunit100 @kubernetes/sig-api-machinery-pr-reviews @kubernetes/sig-scalability-pr-reviews

Commit:
Add NewTimer mockability to Clock

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/40759)
<!-- Reviewable:end -->

ncdc commented:
I see I've got some data races and need to add some additional locks. I've got to step out for a bit but I'll work on that soon. In the meantime I would definitely appreciate any reviews you can provide.

smarterclayton commented on cmd/libs/go2idl/informer-gen/generators/informer.go:
```diff
@@ -134,7 +134,7 @@ type $.type|private$Informer struct {
 
 var typeInformerConstructorInternal = `
-func new$.type|public$Informer(client $.clientSetInterface|raw$, resyncPeriod $.timeDuration|raw$) $.cacheSharedIndexInformer|raw$ {
+func new$.type|public$Informer(client $.clientSetInterface|raw$, resyncCheck, resyncPeriod $.timeDuration|raw$) $.cacheSharedIndexInformer|raw$ {
```
Why do we need two periods?  Why not derive a resyncCheck from the resyncPeriod?

smarterclayton replied:
I'm fairly certain you can spitball an algorithm that does a reasonable job of not letting resync periods go too long and hide that complexity.   Is there an example of a controller that really cares about its resync check period (in a way that isn't exactly the same as all other controllers)?

ncdc commented:
I'll put together an adaptive algorithm. It shouldn't be too much to
implement.

On Tue, Jan 31, 2017 at 10:37 PM Clayton Coleman <notifications@github.com>
wrote:

> *@smarterclayton* commented on this pull request.
> ------------------------------
>
> In cmd/libs/go2idl/informer-gen/generators/informer.go
> <https://github.com/kubernetes/kubernetes/pull/40759>:
>
> > @@ -134,7 +134,7 @@ type $.type|private$Informer struct {
>  `
>
>  var typeInformerConstructorInternal = `
> -func new$.type|public$Informer(client $.clientSetInterface|raw$, resyncPeriod $.timeDuration|raw$) $.cacheSharedIndexInformer|raw$ {
> +func new$.type|public$Informer(client $.clientSetInterface|raw$, resyncCheck, resyncPeriod $.timeDuration|raw$) $.cacheSharedIndexInformer|raw$ {
>
> I'm fairly certain you can spitball an algorithm that does a reasonable
> job of not letting resync periods go too long and hide that complexity. Is
> there an example of a controller that really cares about its resync check
> period (in a way that isn't exactly the same as all other controllers)?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/40759>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAABYqOS9c7PNrgF5tLAeQ3iju8Tgv-Oks5rX_3wgaJpZM4LzKUQ>
> .
>

deads2k commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -118,6 +118,11 @@ type DeltaFIFO struct {
 	// when Replace() or Delete() is called.
 	knownObjects KeyListerGetter
+
+	// ResyncRunning, if set, is invoked at the beginning and end of a resync.
```
give an example of why I would use this?

ncdc replied:
Will do

deads2k commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -407,6 +420,24 @@ func (f *DeltaFIFO) Pop(process PopProcessFunc) (interface{}, error) {
 			f.cond.Wait()
 		}
 		id := f.queue[0]
+		if id == resyncEndKey || id == relistEndKey {
```
I'm suspicious of markers.  Are there comments in here explaining why this was the only solution?

ncdc replied:
I will add them

deads2k commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -428,13 +459,29 @@ func (f *DeltaFIFO) Pop(process PopProcessFunc) (interface{}, error) {
 // after calling this function. f's queue is reset, too; upon return, it
 // will contain the items in the map, in no particular order.
 func (f *DeltaFIFO) Replace(list []interface{}, resourceVersion string) error {
+	if f.RelistRunning != nil {
```
doesn't replace get called on resyncs?  did that change?

ncdc replied:
No, Resync is called on resyncs. Replace is called for relists. But please do doublecheck my research!

deads2k commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -240,14 +267,43 @@ func (s *sharedIndexInformer) GetController() Controller {
+		return check
+	}
+	if desired%check != 0 {
+		// e.g. if resyncPeriod is 5s and resyncCheckPeriod is 3s, we want (floor(5/3) * 3) + 3 = 6
```
This doesn't match my expectation.  If I want 5 and it checks every three, seems like I would just check if my 5 seconds was done each time.  Why try to make them match as a multiple?  It will reduce smoothness, but it seems like its closer to correct.

ncdc replied:
So just remove the modulo check is what you're saying?

deads2k commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -280,45 +345,106 @@ func (s *sharedIndexInformer) HandleDeltas(obj interface{}) error {
+	defer p.listenersLock.RUnlock()
+
+	for _, listener := range p.listeners {
+		listener.resyncRunning(running)
```
doesn't this race with `distribute` since you're changing the value of the `listener` under a read lock instead of a write lock?

ncdc replied:
Yes I need to fix data races. On my todo list.

ncdc replied:
(the listenersLock, though, is to guard against races adding to the listeners slice)

deads2k replied:
> (the listenersLock, though, is to guard against races adding to the listeners slice)

Pretty sure that changing the value of something in one thread that another thing is reading even if its a member of the slice will be a data race.  Thread 2 may never observe the change, which would break this logic.

deads2k commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -419,3 +566,48 @@ func (p *processorListener) run(stopCh <-chan struct{}) {
+// shouldHandleSyncItem returns true if the reflector has started a relisting, or if it has started
+// resyncing and the listener is participating in the current resync.
+func (p *processorListener) shouldHandleSyncItem() bool {
+	return p.relisting || p.resyncing && p.wantSyncItems
```
Wow, that's a lot of plumbing to end up here.  I'm going to think.

deads2k replied:
Ok, if I've read your intent correctly, you're doing this so that a listener doesn't listen to half a resync and think its done syncing.  I think the logic may fail on very slow syncs (see my previous comment), but if that's why, what if we considered changing the deltafifo semantics.  What if doing a resync blocked other actions?  The resync then becomes atomic and whether or not we're currently operating a resync becomes fully discoverable at this layer without any plumbing, right?

ncdc replied:
It's because we need to know when we've popped off the last item from a resync, so that we can ignore future sync items until we hit the next resync period. However, as we discussed on irc, it sounds like making resync atomic would clean up this code significantly. I'll open up a separate pull for that (although it may be a few days before I can get to it).

deads2k commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -280,45 +345,106 @@ func (s *sharedIndexInformer) HandleDeltas(obj interface{}) error {
+	defer p.listenersLock.RUnlock()
+
 	for _, listener := range p.listeners {
+		if isSync && !listener.shouldHandleSyncItem() {
```
This flow doesn't look right to me.  deltafifo has a bunch of items queued on its internal queue, including items from the last resync.  Pop is called which ultimately ends up here with a resync item.  At the same time, Resync is called (which is allowed to queue items async).  The Resync call results in calling `resyncRunning` here, which affects the outcome of `shouldHandleSyncItem`, which ultimately changes the events you were expecting.

deads2k replied:
I'll have to think about whether or not I care.  Resyncs should be safe to duplicate and just annoy the consumers, not break them.

ncdc replied:
You're saying that the shared informer could be a slow consumer of the queue, the queue has items from resync *n*, and while the informer is slowly popping, resync *n+1* happens?

Commit:
Support per-handler resync periods

Add the ability for each event handler of a shared informer to specify
its own resync period. If not specified, a handler will resync at the
informer's default interval.

ncdc commented:
Updated and now based on #40815 (atomic replace in DeltaFIFO).

deads2k commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -240,14 +262,56 @@ func (s *sharedIndexInformer) GetController() Controller {
+		return desired
+	}
+	if check == 0 {
+		glog.Warningf("The specified resyncPeriod %v is invalid because this shared informjer doesn't support resyncing", desired)
```
informer typo

deads2k commented:
lgtm.  This fell out a lot cleaner.

deads2k commented:
Any one else got comments?

ncdc commented:
@k8s-bot kubemark e2e test this

ncdc commented:
@k8s-bot cvm gce e2e test this

`fatal: unable to access 'https://github.com/kubernetes/release/': Could not resolve host: github.com`

ncdc commented:
```
W0208 08:58:12.160] ... calling validate-cluster
I0208 08:58:12.926] Found 2 node(s).
W0208 08:58:13.137] The connection to the server 35.184.71.64 was refused - did you specify the right host or port?
W0208 08:58:13.140] (kubectl failed, will retry 2 times)
W0208 08:58:14.396] The connection to the server 35.184.71.64 was refused - did you specify the right host or port?
W0208 08:58:14.398] (kubectl failed, will retry 1 times)
W0208 08:58:15.561] The connection to the server 35.184.71.64 was refused - did you specify the right host or port?
W0208 08:58:15.564] ('kubectl get nodes' failed, giving up)
```

I don't have access to the GKE master logs so I'm not sure if this was a flake or caused by something in my PR. Looking at the kubelet logs I do have access to, they weren't able to connect to the apiserver, so either the apiserver wasn't running any more or there was some networking issue.

@k8s-bot gci gke e2e test this

deads2k commented:
/lgtm

deads2k commented:
@smarterclayton second look and approval please.

smarterclayton commented:
/approve

ncdc commented:
Just a rebase, re-adding lgtm
